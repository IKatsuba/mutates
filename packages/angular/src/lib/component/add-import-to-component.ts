import type { ClassDeclaration } from 'ts-morph';

import { pushToDecoratorArrayProperty } from '../helpers/push-to-decorator-array-property';

export function addImportToComponent(
  classDeclaration: ClassDeclaration,
  importName: string,
  { unique = false }: { unique?: boolean } = {},
): void {
  pushToDecoratorArrayProperty(classDeclaration, 'Component', 'imports', importName, {
    unique,
    forceToArray: true,
  });
}
