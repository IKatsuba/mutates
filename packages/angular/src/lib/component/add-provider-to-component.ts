import type { ClassDeclaration } from 'ts-morph';

import { pushToDecoratorArrayProperty } from '../helpers/push-to-decorator-array-property';

export function addProviderToComponent(
  classDeclaration: ClassDeclaration,
  provider: string,
  { unique = false }: { unique?: boolean } = {},
): void {
  pushToDecoratorArrayProperty(classDeclaration, 'Component', 'providers', provider, {
    unique,
    forceToArray: true,
  });
}
