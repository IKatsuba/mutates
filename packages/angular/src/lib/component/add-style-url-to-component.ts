import type { ClassDeclaration } from 'ts-morph';

import { pushToDecoratorArrayProperty } from '../helpers/push-to-decorator-array-property';

export function addStyleUrlToComponent(
  classDeclaration: ClassDeclaration,
  styleUrl: string,
  { unique = false }: { unique?: boolean } = {},
): void {
  pushToDecoratorArrayProperty(classDeclaration, 'Component', 'styleUrls', styleUrl, {
    unique,
  });
}
