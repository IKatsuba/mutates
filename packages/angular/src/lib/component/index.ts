export * from './add-import-to-component';
export * from './add-provider-to-component';
export * from './add-style-url-to-component';
