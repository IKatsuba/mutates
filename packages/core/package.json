{
  "name": "@mutates/core",
  "version": "0.0.0-development",
  "keywords": [
    "typescript",
    "ast",
    "mutations"
  ],
  "homepage": "https://github.com/IKatsuba/mutates",
  "repository": "https://github.com/IKatsuba/mutates",
  "license": "Apache-2.0",
  "contributors": [
    "Igor Katsuba <igor@katsuba.dev>"
  ],
  "type": "commonjs",
  "main": "./src/index.js",
  "typings": "./src/index.d.ts",
  "dependencies": {
    "minimatch": "9.0.3",
    "ts-morph": "^22.0.0",
    "tslib": "^2.3.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "authors": [
    "Igor Katsuba <igor@katsuba.dev>"
  ]
}
