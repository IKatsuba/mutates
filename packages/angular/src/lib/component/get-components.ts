import type { ClassDeclaration } from 'ts-morph';

import { getClasses, Pattern, Query, StructureType } from '@mutates/core';

export function getComponents(
  pattern: Pattern,
  query?: Query<Omit<StructureType<ClassDeclaration>, 'kind'>>,
): ClassDeclaration[] {
  return getClasses(pattern, query).filter(isComponent);
}

export function isComponent(declaration: ClassDeclaration): boolean {
  return !!declaration.getDecorator('Component');
}
