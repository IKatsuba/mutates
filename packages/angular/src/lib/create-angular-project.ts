import { Tree } from '@angular-devkit/schematics';
import { ProjectOptions } from 'ts-morph';

import { createProject } from '@mutates/core';

import { NgTreeFileSystem } from './ng-tree-file-system';

export function createAngularProject(tree: Tree, options?: Omit<ProjectOptions, 'fileSystem'>) {
  return createProject(new NgTreeFileSystem(tree), options);
}
