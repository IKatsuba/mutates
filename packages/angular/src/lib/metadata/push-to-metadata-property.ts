import { ClassDeclaration, SyntaxKind } from 'ts-morph';

import { getObjectProperties, pushToArrayIfNotExists } from '@mutates/core';

import { MetadataProperty } from '../types/metadata-property';
import { getMetadata } from './get-metadata';

export function pushToMetadataProperty<T extends MetadataProperty>(
  klass: ClassDeclaration,
  property: T,
  values: string[],
): void {
  const [metadata] = getMetadata(klass);

  const prop = getObjectProperties(metadata, {
    name: property as string,
  }).at(0);

  if (prop) {
    const initializer = prop.getInitializerIfKind(SyntaxKind.ArrayLiteralExpression);

    if (initializer) {
      pushToArrayIfNotExists(initializer, ...values);
    }
  } else {
    metadata.addPropertyAssignment({
      name: property as string,
      initializer: `[${values.join(', ')}]`,
    });
  }
}
