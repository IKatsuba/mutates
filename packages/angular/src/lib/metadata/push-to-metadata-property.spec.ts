import { Tree } from '@angular-devkit/schematics';

import { createSourceFile, readFileSync, saveProject } from '@mutates/core';

import { createTestingTree } from '../../testing';
import { getComponents } from '../component';
import { createAngularProject } from '../create-angular-project';
import { pushToMetadataProperty } from './push-to-metadata-property';

describe('pushToMetadataProperty', () => {
  let host: Tree;
  beforeEach(() => {
    host = createTestingTree();

    createAngularProject(host);
  });

  it('should add providers to component', () => {
    createSourceFile(
      'src/main.ts',
      `
@Component({}) class AppComponent {}
`,
    );

    pushToMetadataProperty(getComponents().at(0)!, 'providers', ['AppService']);

    saveProject();

    expect(readFileSync('src/main.ts')).matchSnapshot();
  });
});
