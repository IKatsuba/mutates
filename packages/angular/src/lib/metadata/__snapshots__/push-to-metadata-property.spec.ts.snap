// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`pushToMetadataProperty > should add providers to component 1`] = `
"
@Component({
    providers: [AppService]
}) class AppComponent {}
"
`;
